packages/server/src/lib/grants-ingest.js

```diff
@@ -0,0 +1,156 @@
+    if (messages.length === 0) {
+        console.log('Empty message batch received from SQS');
+    }
+    return messages;
+}
+
+function sqsMessageToGrant(jsonBody) {
+    const messageData = JSON.parse(jsonBody);
+    return {
+        status: 'inbox',
+        grant_id: messageData.OpportunityId || messageData.grant_id,
+        grant_number: messageData.OpportunityNumber,
+        agency_code: messageData.AgencyCode,
+        award_ceiling: (messageData.AwardCeiling && parseInt(messageData.AwardCeiling, 10))
+            ? parseInt(messageData.AwardCeiling, 10) : undefined,
+        award_floor: (messageData.AwardFloor && parseInt(messageData.AwardFloor, 10))
+            ? parseInt(messageData.AwardFloor, 10) : undefined,
+        cost_sharing: messageData.CostSharingOrMatchingRequirement ? 'Yes' : 'No',
+        title: messageData.OpportunityTitle,
+        cfda_list: (messageData.CFDANumbers || []).join(', '),
+        open_date: normalizeDateString(messageData.PostDate),
+        close_date: normalizeDateString(messageData.CloseDate || '2100-01-01'),
+        notes: 'auto-inserted by script',
+        search_terms: '[in title/desc]+',
+        reviewer_name: 'none',
+        opportunity_category: opportunityCategoryMap[messageData.OpportunityCategory],
+        description: messageData.Description,
+        eligibility_codes: (messageData.EligibleApplicants || []).join(' '),
+        opportunity_status: 'posted',
+        raw_body: JSON.stringify(messageData),
+    };
+}
+
+async function upsertGrant(knex, grant) {
+    await knex('grants')
+        .insert(grant)
+        .onConflict('grant_id')
+        .merge({ ...grant, ...{ updated_at: 'now' } })
+        .returning('grant_id');
+}
+
+async function deleteMessage(sqs, queueUrl, receiptHandle) {
+    const command = new DeleteMessageCommand({
+        QueueUrl: queueUrl,
+        ReceiptHandle: receiptHandle,
+    });
+    await sqs.send(command);
+}
+
+/**
+ * processMessages Saves a batch of SQS messages containing JSON grant data sent
+ * from the grants-ingest service to the database. Existing database records that match
+ * a message's grant identifier are updated. After each message is processed, it is
+ * deleted from the SQS queue.
+ *
+ * Any errors related to parsing or saving are logged and do not prevent further processing.
+ * Errors interacting with SQS are fatal.
+ *
+ * @param { import('knex').Knex } knex Database client for persisting grants.
+ * @param { import('@aws-sdk/client-sqs').SQSClient } sqs AWS SDK client used to issue commands to the SQS API.
+ * @param { string } queueUrl The URL identifying the queue to poll.
+ * @param { Array[import('@aws-sdk/client-sqs').Message] } messages Messages to process from SQS.
+ */
+async function processMessages(knex, sqs, queueUrl, messages) {
+    let grantParseErrorCount = 0;
+    let grantSaveSuccessCount = 0;
+    let grantSaveErrorCount = 0;
+
+    return Promise.all(messages.map(async (message) => {
```

Comment:
Note: Not critical for this PR, but something to consider as we build on it.

I would consider using `asyncBatch` for this to ensure we do not get the same memory overload issues we have seen in the past. It looks like we already cap the messages length in line 39 so it should be OK here. Although having the limit enforced at the point where we are calling these functions might make it more explicit.

Reply:
I did consider using `asyncBatch` for this, but ultimately decided it didn't seem necessary given that the `ReceiveMessage` behavior already effectively limits the pool to a concurrency level of <= 10. 

If the main loop was engaging in more of a "fire and forget" pattern where it polled for up to 10 messages, sent them to a separate pool of async workers, and then immediately started polling again (without waiting for the previous 10 messages to process), we would definitely want to make use of `asyncBatch` to enforce an upper bound on the number of workers.